test/common/http/http2/http2_connection_fuzz_test.cc

```diff
@@ -0,0 +1,307 @@
+                     std::string& payload) {
+#define MAX_HDRS 32
+  nghttp2_nv h2hdrs[MAX_HEADERS];
+  uint8_t buf[MAX_HD_TABLE_BUF_SIZE];
+
+  unsigned n_headers = fragment.headers().size();
+  if (n_headers > MAX_HEADERS)
+    n_headers = MAX_HEADERS;
+  auto headers = fragment.headers();
+  for (unsigned i = 0; i < n_headers; i++) {
+    auto hv = headers.at(i);
+    size_t keylen = hv.key().size();
+    uint8_t* key = reinterpret_cast<uint8_t*>(alloca(keylen));
+    std::memcpy(key, hv.key().data(), keylen);
+
+    size_t valuelen = hv.value().size();
+    uint8_t* value = reinterpret_cast<uint8_t*>(alloca(valuelen));
+    std::memcpy(value, hv.value().data(), valuelen);
+
+    h2hdrs[i].name = key;
+    h2hdrs[i].namelen = keylen;
+    h2hdrs[i].value = value;
+    h2hdrs[i].valuelen = valuelen;
+    h2hdrs[i].flags = 0;
+  }
+  ssize_t len = nghttp2_hd_deflate_hd(deflater, buf, sizeof(buf), h2hdrs, n_headers);
+  if (len > 0)
+    payload.append(reinterpret_cast<char*>(buf), len);
+}
+
+Http2Frame pb_to_h2_frame(nghttp2_hd_deflater* deflater,
+                          const test::common::http::http2::Http2FrameOrJunk& pb_frame) {
+  if (pb_frame.has_junk()) {
+    // Junk frame
+    auto junk = pb_frame.junk();
+    absl::string_view contents = junk.data();
+    return Http2Frame(contents);
+  } else if (pb_frame.has_h2frame()) {
+    // Need to encode headers
+    auto h2frame = pb_frame.h2frame();
+    uint32_t streamid = h2frame.streamid();
+    uint8_t flags = static_cast<uint8_t>(h2frame.flags() & 0xff);
+    bool use_padding = flags & Http2Frame::PADDED;
+    uint8_t type;
+    std::string payload;
+    if (h2frame.has_data()) {
+      type = 0;
+      auto f = h2frame.data();
+      payload.append(f.data());
+    } else if (h2frame.has_headers()) {
+      type = 1;
+      auto f = h2frame.headers();
+      uint8_t padding_len = static_cast<uint8_t>(f.padding().size() & 0xff);
+      if (use_padding) {
+        payload.push_back(static_cast<char>(padding_len));
+      }
+
+      uint32_t stream_dependency = f.stream_dependency();
+      payload.append(reinterpret_cast<char*>(&stream_dependency), sizeof(stream_dependency));
+
+      if (f.weight() > 0 && f.weight() <= 256) {
+        uint8_t weight = static_cast<uint8_t>((f.weight() - 1) & 0xff);
+        payload.append(reinterpret_cast<char*>(&weight), sizeof(weight));
+      }
+
+      deflate_headers(deflater, f.headers(), payload);
+      if (use_padding) {
+        payload.append(f.padding().data(), padding_len);
+      }
+    } else if (h2frame.has_priority()) {
+      type = 2;
+      auto f = h2frame.priority();
+      uint32_t stream_dependency = f.stream_dependency();
+      payload.append(reinterpret_cast<char*>(&stream_dependency), sizeof(stream_dependency));
+      payload.push_back(static_cast<char>(f.weight() & 0xff));
+    } else if (h2frame.has_rst()) {
+      type = 3;
+      auto f = h2frame.rst();
+      uint32_t error_code = f.error_code();
```

Comment:
Do we need calls to htonl() when serializing integers? (Here and below.)

Reply:
Not necessarily, the fuzzer can generate integers in either byte order. However, it might make a crash case more (human-)readable.